engine/userguide/eng-image/multistage-build.md

```diff
@@ -0,0 +1,158 @@
+---
+description: Keeping your images small with multi-stage images
+keywords: images, containers, best practices
+title: Use multi-stage builds
+---
+
+{% include edge_only.md section="page" %}
+
+Multi-stage builds are a new feature in Docker 17.05, and they will be exciting
+to anyone who has struggled to optimize Dockerfiles while keeping them easy to
+read and maintain.
+
+> **Acknowledgment**:
+> Special thanks to [Alex Ellis](https://twitter.com/alexellisuk) for granting
+> permission to use his blog post
+> [Builder pattern vs. Multi-stage builds in Docker](http://blog.alexellis.io/mutli-stage-docker-builds/)
+> as the basis of the examples below.
+
+## Before multi-stage builds
+
+One of the most challenging things about building images is keeping the image
+size down. Each instruction in the Dockerfile adds a layer to the image, and you
+need to remember to clean up any artifacts you don't need before moving on to
+the next layer. To write a really efficient Dockerfile, you have traditionally
+needed to employ shell tricks and other logic to keep the layers as small as
+possible and to ensure that each layer has the artifacts it needs from the
+previous layer and nothing else.
+
+It was actually very common to have one Dockerfile to use for development (which
+contained everything needed to build your application), and a slimmed-down one
+to use for production, which only contained your application and exactly what
+was needed to run it. This has been referred to as the "builder
+pattern". Maintaining two Dockerfiles is not ideal.
+
+Here's an example of a `Dockerfile.build` and `Dockerfile` which adhere to the
+builder pattern above:
+
+**`Dockerfile.build`**:
+
+```conf
+FROM golang:1.7.3
+WORKDIR /go/src/github.com/alexellis/href-counter/
+RUN go get -d -v golang.org/x/net/html  
+COPY app.go    .
+RUN go get -d -v golang.org/x/net/html \
+  && CGO_ENABLED=0 GOOS=linux go build -a -installsuffix cgo -o app .
+```
+
+Notice that this example also artificially compresses two `RUN` commands together
+using the Bash `&&` operator, to avoid creating an additional layer in the image.
+This is failure-prone and hard to maintain. It's easy to insert another command
+and forget to continue the line using the `\` character, for example.
+
+**`Dockerfile`**:
+
+```conf
+FROM alpine:latest  
+RUN apk --no-cache add ca-certificates
+WORKDIR /root/
+COPY app .
+CMD ["./app"]  
+```
+
+**`build.sh`**:
+
+```bash
+#!/bin/sh
+echo Building alexellis2/href-counter:build
+
+docker build --build-arg https_proxy=$https_proxy --build-arg http_proxy=$http_proxy \  
+    -t alexellis2/href-counter:build . -f Dockerfile.build
+
+docker create --name extract alexellis2/href-counter:build  
+docker cp extract:/go/src/github.com/alexellis/href-counter/app ./app  
+docker rm -f extract
+
+echo Building alexellis2/href-counter:latest
+
+docker build --no-cache -t alexellis2/href-counter:latest .  
```

Comment:
@mstanleyjones let's add an `rm ./app` here so that things are belt and braces. I think I just maintain a .gitignore.